### PR TITLE
fix(grainfmt): Handle comments within lists

### LINKED ISSUE #1417: GrainFmt: Grain Formatter Removing Comments
When grain formats lists containing comments it deletes them as can be seen by the before and after below.
Before formatting ```
let test = [
  1, // test
  2, // test
  3, // test
``` after formatting ```
let test = [ 1, 2, 3]
```

### PR DESCRIPTION
**Pull request by marcusroberts**
Comments within lists were being discarded.

Fixes #1417

### CODE REVIEW

**Comment by marcusroberts** [comment]
Just checking why the stdlib format has changed, it shouldn't have had an impact

**Comment by marcusroberts** [comment]
Switching to draft, as the format has changed for compiler/test/stdlib/hash.test.gr as a list of records now breaks over line.

**Comment by marcusroberts** [comment]
All good now, but discovered #1428 needs more work, but this can go in independently.

File: compiler/src/formatting/format.re [hunk 1]
```diff
@@ -1065,41 +1065,150 @@ and resugar_list =
...
 
+  // We have to compose this list by hand because of the complexity of if a list item
+  // is followed by a comment, the comma must come before the comment.
+  // It also impacts how we force a new line for a line ending comment at the end of a list
+  // without introducing an extra blank line when bringing the indentation back in again
+
+  let last_line_breaks_for_comments = ref(false);
+  let items_length = List.length(items);
+  let list_items =
+    Doc.concat(
```
Comment (phated) [inline compiler/src/formatting/format.re:1152]: Can this concat be moved below (line 1199) instead? It feels weird here and makes the nesting harder to read

File: compiler/src/formatting/format.re [hunk 2]
```diff
@@ -1065,41 +1065,150 @@ and resugar_list =
...
+      List.mapi(
+        (i, pair) => {
+          let (item, item_comments) = pair;
```
Comment (phated) [inline compiler/src/formatting/format.re:1155]: Is `pair` used anywhere or should this destructure happen in the function definition?

**Review by phated — COMMENTED** [review]
A couple comments. This fix was a bit harder to follow that my other approvals (I think it is due to so many `Doc.concat` combinations) but the tests look good.

**Comment by marcusroberts** [comment]
Thanks for the great comments, fixed as suggested.

**Review by ospencer — APPROVED** [review]
Looks great!

**Review by phated — APPROVED** [review]

**Pull request merged**